### PR TITLE
WIP: Cmake port

### PR DESCRIPTION
**Pull request by TomasPuverle**
This is a WIP but it's already useful.

If you know cmake, this should be straightforward to use.  I will update the README once closer to being finished.  

For the uninitiated, see [here](https://cliutils.gitlab.io/modern-cmake/chapters/intro/running.html) or just follow the below:

1. cmake >= 3.15 required
2. In your srcdir, run `cmake -S . -B <build_dir>` or `cmake -S . -B <build_dir> -D<option>=<value>` if you want to control settings (documented in the root CMakeFile.txt).  You can also use `ccmake <build_dir>` after to see which options are set.
3. The run `cmake --build <build_dir> [-j16]` and souffle will build
4. To run tests, `cd <build_dir>; ctest [-j16]` or even just `cmake --build <build_dir> [-j16] --target test`

I've made sure the compilation flags are the same on Linux 32/64.  You can run both cmake and ctest with `--verbose` to see compile lines or test executions.
Not all the tests are done yet - all the unit tests are included but not all of the positive/negative souffle tests are done.

A few notes on improvements:

1. It's not automake ;).  No need for `bootstrap/configure`
2. In general, I think the make structure is cleaner.  Right now, the source layout is still automake compatible since I want the two to live side-by-side but there is some room for improvement
3. Clear targets which can be built.  See `cmake --build <build_dir>  --target help` to see what's available
4. Testing is improved; all tests are named (so they can be run by name using `ctest -R/--tests-regex` or by label, e.g. `cmake -L/--label-regex`.  It's now possible to e.g. run just the "interpreted" tests so that testing turnaround is improved.  `--rerun-failed` is your friend, too.
5. I broke out each test into four phases - setup, test, compare stderr/out, compare out files.  That way it's easier to spot which part of the test failed.  That's why you will now see about 4x more tests.
6. The build is now out-of-source by default and it's possible/easier to have e.g. a release and a debug build side-by-side.  Just configure into two separate build dirs. 
7. Support from editors (including Intellisense) such as VS Code, CLion and Visual Studio.

There are also performance advantages.  On my machine, there are the build/test times:

automake, with -j4:
clean build: 3:12.24 total

cmake, with -j4:
clean build: 1:55.35 total (This actually does more since it also builds the unit tests)

There's still some work to be done; there is a list of TODO's in the root CMakeFile.txt.

I would appreciate it if someone could give this a spin on Ubuntu/MacOS to see if there are any issues.

### CODE REVIEW

**Comment by XiaowenHu96** [comment]
Hi Tomas, I tested it on two of our services, but couldn't build them. (I'm fairly new to cmake so it also might be a misconfigure).
Here is what I did on a Ubuntu service (cmake 3.16.3):
1. Make a clean clone.
2. `cd souffle && mkdir build && cmake -S . -B build`

I got following output:
```
-- Building souffle version 2.0.2-993-g49da3f0b3
-- The CXX compiler identification is GNU 10.2.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Build Type: Release
-- Found Curses: /usr/lib/x86_64-linux-gnu/libcurses.so
-- Found FLEX: /usr/bin/flex (found version "2.6.4")
-- Found BISON: /usr/bin/bison (found suitable version "3.7", minimum required is "3.0.4")
-- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11")
-- Found SQLite3: /usr/include (found version "3.33.0")
-- Found LibFFI: /usr/include/x86_64-linux-gnu
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Found Threads: TRUE
-- Found OpenMP_CXX: -fopenmp (found version "4.5")
-- Found OpenMP: TRUE (found version "4.5") found components: CXX
CMake Error: File /home/xihu5895/souffle-cmake/cmake/config.h.in does not exist.
CMake Error at CMakeLists.txt:134 (configure_file):
  configure_file Problem configuring file

-- Configuring incomplete, errors occurred!
See also "/home/xihu5895/souffle-cmake/build/CMakeFiles/CMakeOutput.log".
See also "/home/xihu5895/souffle-cmake/build/CMakeFiles/CMakeError.log".
```

From the CMakeError:
```
Performing C++ SOURCE FILE Test CMAKE_HAVE_LIBC_PTHREAD failed with the following output:
Change Dir: /home/xihu5895/souffle-cmake/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake cmTC_8a95e/fast && /usr/bin/gmake -f CMakeFiles/cmTC_8a95e.dir/build.make CMakeFiles/cmTC_8a95e.dir/build
gmake[1]: Entering directory '/home/xihu5895/souffle-cmake/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_8a95e.dir/src.cxx.o
/usr/bin/c++    -DCMAKE_HAVE_LIBC_PTHREAD   -std=c++17 -o CMakeFiles/cmTC_8a95e.dir/src.cxx.o -c /home/xihu5895/souffle-cmake/build/CMakeFiles/CMakeTmp/src.cxx
Linking CXX executable cmTC_8a95e
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_8a95e.dir/link.txt --verbose=1
/usr/bin/c++   -DCMAKE_HAVE_LIBC_PTHREAD    CMakeFiles/cmTC_8a95e.dir/src.cxx.o  -o cmTC_8a95e
/usr/bin/ld: CMakeFiles/cmTC_8a95e.dir/src.cxx.o: in function `main':
src.cxx:(.text+0x46): undefined reference to `pthread_create'
/usr/bin/ld: src.cxx:(.text+0x52): undefined reference to `pthread_detach'
/usr/bin/ld: src.cxx:(.text+0x63): undefined reference to `pthread_join'
collect2: error: ld returned 1 exit status
gmake[1]: *** [CMakeFiles/cmTC_8a95e.dir/build.make:87: cmTC_8a95e] Error 1
gmake[1]: Leaving directory '/home/xihu5895/souffle-cmake/build/CMakeFiles/CMakeTmp'
gmake: *** [Makefile:121: cmTC_8a95e/fast] Error 2


Source file was:
#include <pthread.h>

void* test_func(void* data)
{
  return data;
}

int main(void)
{
  pthread_t thread;
  pthread_create(&thread, NULL, test_func, NULL);
  pthread_detach(thread);
  pthread_join(thread, NULL);
  pthread_atfork(NULL, NULL, NULL);
  pthread_exit(NULL);

  return 0;
}
```

I got a similar result on another service running Fedora (Cmake 3.17.4)

**Comment by TomasPuverle** [comment]
Hi @XiaowenHu96,

Thanks for your help!  It turns out that there was a pattern of `config.*` in .gitignore, which meat that my `config.h.in` file didn't get added.  I have fixed this now, please have a try.

Please note that the tests are still a little flaky, I am looking into it.  Please let me know if you find other issues.

Best,

Tom

**Comment by TomasPuverle** [comment]
It turns out the flakyness I'm seeing is because "example" is not usually part of the tests. There are examples that actually fail!

**Comment by TomasPuverle** [comment]
@XiaowenHu96 Have you had a chance to try the build?  Did it work for you this time?

As an aside, it seems this work is coming at a good time:  This autotools thread seems somewhat ominous: [https://lists.gnu.org/archive/html/autoconf/2021-01/msg00049.html](https://lists.gnu.org/archive/html/autoconf/2021-01/msg00049.html)

**Comment by XiaowenHu96** [comment]
@TomasPuverle I just tested the build again on our servers, everything looks good to me!
I'll see if I can find a MacOS to test on.

**Comment by TomasPuverle** [comment]
Thanks!  I'll try to get the rest of the build working soon.  The next part of the tests are a little "trickier" especially because there's some additional work to support Windows builds.  I may punt on that for now and fix as needed, we'll see.

**Comment by TomasPuverle** [comment]
@XiaowenHu96 @b-scholz 

Today/tonight was make/shell script hell but I think swig is done.  If you could try on Ubuntu/MacOS I would appreciate it.

Make sure to enable swig, otherwise the tests won't run (as they should).  When configuring, run e.g.

`cmake -S . -B build -DSOUFFLE_SWIG` or just `cmake -S . -B build -DSOUFFLE_SWIG_PYTHON` (or JAVA) to enable swig just for one language.

**Comment by TomasPuverle** [comment]
I'm not sure if anyone on this thread is using VS Code, but if you install the Microsoft C++ extension, it has really nice integration with CMake.  You can work/build/debug/change build variant etc. directly from the editor.  Even running the tests with a click of button.

Even perhaps nicer, it has really neat support for containers.  We could just add a .devcontainer directory to the repo and - directly in VSCode - you can just enter the github url and it will:

1) clone
2) build the container
3) launch a remote session in the container
4) install the required extensions (such as C++ and souffle) 

Basically create a full dev env/IDE with all dependencies in a single step!  Here's an example video which show it in action: [YouTube](https://www.youtube.com/watch?v=Uvf2FVS1F8k)

**Comment by b-scholz** [comment]
Great progress!

**Comment by XiaowenHu96** [comment]
@TomasPuverle 
I just noticed test cases are failing randomly with `-j` flag, can it be a race condition?

**Comment by XiaowenHu96** [comment]
Also, on our Ubuntu server, those swig tests failed even without `-j` flag. I don't know anything about SWIG, is there anything I can do on my side to help you identify the issue?
```
        2395 - swig/python/flights_compare_csv (Failed)
        2401 - swig/java/flights_compare_csv (Failed)
        2406 - swig/python/insert_for_compare_csv (Failed)
        2412 - swig/java/insert_for_compare_csv (Failed)
        2417 - swig/python/movies_compare_csv (Failed)
        2423 - swig/java/movies_compare_csv (Failed)
        2428 - swig/python/paths_compare_csv (Failed)
        2434 - swig/java/paths_compare_csv (Failed)
```

**Comment by TomasPuverle** [comment]
@XiaowenHu96 Thanks so much for trying.  It could be a race but I've tried to ensure the tests have proper dependencies between them.  
Could you please send me the commands how you configured cmake?  Also, try running ctest (from the build directory) as `ctest -L swig --verbose`.  That should only run the swig tests and give more output.
Do you happen to use a chat client of some kind (like Slack)?  It may be faster than github to figure out what's wrong.  I tried joining Zulip but nobody's there.

**Comment by XiaowenHu96** [comment]
@TomasPuverle 
My slack: xihu5895@uni.sydney.edu.au
or if Skype works better for you:
xihu0208@gmail.com

**Comment by TomasPuverle** [comment]
Thanks for your help @XiaowenHu96 . Hopefully this fixes the issues you were seeing.
Another reason to port these scripts from shell to Python :snake: :smiley:

**Comment by TomasPuverle** [comment]
@b-scholz 

Question:  In tests/interface.at, do you know why the following code/comment exists?  Starting at line 23:

 AT_CHECK(["$SOUFFLE" -D- -o $1 -F FACTS PROGRAM 1>TESTNAME.out 2>TESTNAME.err], [0])
    **# remove executable and re-build it from scratch**
    AT_CHECK([rm $1 2>>TESTNAME.err],[0])
    AT_CHECK([$CXX "-I$SOUFFLE_INC" $CXXFLAGS $CPPFLAGS -D__EMBEDDED_SOUFFLE__ -o $1 TESTDIR/driver.cpp $1.cpp $LIBS $LDFLAGS -L$TESTDIR 2>>TESTNAME.err],[0])

Why/what's the need to remove the executable and rebuild it from scratch?  Also, why is it using a direct invocation of the compiler rather than souffle-compile?

**Comment by TomasPuverle** [comment]
@b-scholz Hi Bernhard, ping on the above.  Thanks!

**Comment by darth-tytus** [comment]
I've tested the build on MacOS using `CC=gcc-10 && CXX=g++-10 cmake -S . -B build` as a compile command.

I've had a problem with discovering libffi, which I was able to solve with:
```diff
 # Look for the necessary header
-find_path(LIBFFI_INCLUDE_DIR NAMES ffi.h)
+find_path(LIBFFI_INCLUDE_DIR NAMES ffi.h
+  PATHS /usr /usr/local /opt/local /usr/include/ffi
+  PATH_SUFFIXES include include/ffi include/x86_64-linux-gnu x86_64-linux-gnu)
 mark_as_advanced(LIBFFI_INCLUDE_DIR)
```
Taken from https://github.com/cmu-db/peloton/blob/master/cmake/Modules/FindLibffi.cmake

Some of the tests are failing, this is probably due to some misconfiguration on my site, since the failed-tests sequence mostly follows arithmetic progression `22, 30, 38, 46...`

**Comment by TomasPuverle** [comment]
@darth-tytus I updated the branch as per your suggestion, thanks!  LMK if it works "out of the box" now.

**Comment by darth-tytus** [comment]
@TomasPuverle Yes, thank you :)

**Comment by b-scholz** [comment]
Sorry for the slow feedback. I have been looking into the build process / Jenkins. There are a lot of scripts that use autoconf/automake. We need to change them. 
A Jenkins script looks as follows:

Master-distcheck:
```
chmod -f -R +w * || true
#source scl_source enable devtoolset-7 && git clean -ffxd && ./bootstrap && ./configure --prefix=/tmp/souffle && AM_MAKEFLAGS=-j4 TESTSUITEFLAGS=-j4 make distcheck -j4 && make install && /tmp/souffle/bin/souffle -c tests/example/metro/metro.dl -D- && make uninstall
docker run --rm --mount  src="$(pwd)",target=/souffle,type=bind -u 1001 ubuntu_lts \
sh -c "cd /souffle \
&& ./bootstrap \
&& ./configure --prefix=/tmp/souffle \
&& AM_MAKEFLAGS=-j4 TESTSUITEFLAGS=-j4 make distcheck -j4 \
&& make install -j4 \
&& /tmp/souffle/bin/souffle -c tests/example/metro/metro.dl -D- \
&& make uninstall"
```
Master-Fedora:
```
#./bootstrap && ./configure && make -j2 && TESTSUITEFLAGS=-j2 make check -j2

docker run --rm -t --mount  src="$(pwd)",target=/souffle,type=bind -u 1001 fedora_lts sh -c "cd souffle \
&& ./bootstrap \
&& ./configure \
&& make -j4 \
&& TESTSUITEFLAGS=-j4 make check -j4 || (.travis/after_failure.sh && false)"
```

Master-memory
```
#20201207: remove --enable-debug

docker run --rm -t --mount  src="$(pwd)",target=/souffle,type=bind -u 1001 --cap-add SYS_PTRACE ubuntu_dev \
sh -c "cd /souffle && ./bootstrap \
&& ./configure --enable-sanitise-memory \
&& make -j4 \
&& TESTSUITEFLAGS=-j4 SOUFFLE_CATEGORY=Syntactic,Semantic,FastEvaluation,Profile,Interface,Provenance make check -j4 \
|| (.travis/after_failure.sh && false)"
```

Master-OSX:
```
export CC=/usr/local/bin/gcc-10
export CXX=/usr/local/bin/g++-10
export PATH=/usr/local/opt/bison/bin:/usr/local/bin:$PATH
export PKG_CONFIG_PATH="/usr/local/opt/libffi/lib/pkgconfig:$PATH"
export TERM=xterm

./bootstrap
./configure
make -j2
TESTSUITEFLAGS=-j2 make check -j2 || (.travis/after_failure.sh && false)
```

```

Master-Package-Debian:
```
# 10/04/2020 Depedencies differ between Ubuntu Focal and previous versions so we need a package for each.
# If anything more is needed, the repo setup should be revisited - more repos or more directories.

docker run -e BINTRAY_API_KEY --rm -t --mount  src="$(pwd)",target=/souffle,type=bind -u 1001 debian_lts \
sh -c "set -x;cd /souffle \
&& ./bootstrap \
&& ./configure --enable-host-packaging \
&& make -j2 package \
&& . /etc/os-release \
&& cd packaging \
&& rename \"s/-1_/-1_\${VERSION_CODENAME}_/\" *\.deb \
&& export DEBNAME=\$(ls souffle_*\.deb) \
&& curl -X PUT -T \$DEBNAME -usouffle-lang:\$BINTRAY_API_KEY \
\"https://api.bintray.com/content/souffle-lang/deb-unstable/souffle/\$(git describe --tags --always)/pool/main/s/souffle/\$DEBNAME;deb_distribution=\$VERSION_CODENAME;deb_component=main;deb_architecture=amd64;publish=1\" \
&& if [ \$(git describe --tags) = \$(git describe --tags --abbrev=0) ] \
; then curl -X PUT -T \$DEBNAME -usouffle-lang:\$BINTRAY_API_KEY \
\"https://api.bintray.com/content/souffle-lang/deb/souffle/\$(git describe --tags --always)/pool/main/s/souffle/\$DEBNAME;deb_distribution=\$VERSION_CODENAME;deb_component=main;deb_architecture=amd64;publish=1\" \
;fi"
```







```
./bootstrap
./configure
make -j24
ln -s src/include include
# run some benchmarks
cd ../../benchmarks/benchmarks
./big_benchmark.sh --instance=/home/souffle-lang/workspace/Master-Performance/src/souffle --outdir=/home/souffle-lang/workspace/Master-Performance/data
```

**Comment by b-scholz** [comment]
There are many more...

I wonder how we should do this. Perhaps we accept first the CMAKE build but keep the autoconf/automake.  We switch the easy Jenkins scripts first and slowly convert the rest. 

Only if the whole Jenkins systems is up and running we can remove the autoconf/automake scripts.

**Comment by TomasPuverle** [comment]
Hi @b-scholz,

Thanks for the follow-up.  Right now, the any of the jobs that just compile and run the tests (something like this)

```
./bootstrap
./configure
make -j2
TESTSUITEFLAGS=-j2 make check -j2 
```

are almost ready to go.  Another thing to consider would be to setup Github actions, which could, at least during the port, provide the additional testing capacity of automake/cmake side-by-side.  Once the two builds are functionally equivalent, the automake build can be disabled, deprecated and removed.

The reason why I say "almost ready" is because I've not ported the "interface" and "provenance" tests yet.  I'm still need some clarification/guidance as per my question above.

Also, CMake doesn't really have a concept of a "distcheck" so we should discuss whether or not it's still needed, or what to do instead.
 
Thanks,

Tom

**Comment by b-scholz** [comment]
I need to check what still goes wrong with the interface. I hope with the compiler-script fix we can switch to cmake soon.

**Comment by b-scholz** [comment]
I had a look today and noticed that cmake causes a problem in OS X. I made a PR for your branch. 

Also, the OPENMP installation should be conditional in OS X since the default clang/llvm system does not ship with OPENMP (NB: this works only with the brew g++ compiler).

**Comment by TomasPuverle** [comment]
Thanks @b-scholz.  Feel free to update this branch (I can do drive-by reviews if you like).  I probably won't have a lot of time for a few more weeks to return to this.
Most important part right now is to ensure that new tests added under automake are also integrated under the new build system.

**Comment by b-scholz** [comment]
> @b-scholz
> 
> Question: In tests/interface.at, do you know why the following code/comment exists? Starting at line 23:
> 
> AT_CHECK(["$SOUFFLE" -D- -o $1 -F FACTS PROGRAM 1>TESTNAME.out 2>TESTNAME.err], [0])
> **# remove executable and re-build it from scratch**
> AT_CHECK([rm $1 2>>TESTNAME.err],[0])
> AT_CHECK([$CXX "-I$SOUFFLE_INC" $CXXFLAGS $CPPFLAGS -D__EMBEDDED_SOUFFLE__ -o $1 TESTDIR/driver.cpp $1.cpp $LIBS $LDFLAGS -L$TESTDIR 2>>TESTNAME.err],[0])
> 
> Why/what's the need to remove the executable and rebuild it from scratch? Also, why is it using a direct invocation of the compiler rather than souffle-compile?


This is related to the days when we did not have the `-g <filename>` flag. We create an executable and as a side-effect, the CPP file is created. We delete the executable and the error file and continue re-compiling the CPP file as a library. 
This is of course outdated and we should use the `-g <filename>` instead.

**Comment by b-scholz** [comment]
We moved to PR #1900.

**Comment by b-scholz** [comment]
This PR is related to issue #1754.
